tests/explicit-nulls/neg/flow6.scala

```diff
@@ -1,70 +1,50 @@
-
-// Test that flow inference behaves soundly within blocks.
-// This means that flow facts are propagated to ValDefs, but not
-// to DefDefs or lazy vals or implicit vals.
-// The reason is that forward references are allowed for `defs`,
-// so when we see
-// ```
-// f
-// if (x == null) return "foo"
-// def f = x.length
-// ```
-// we can't assume that f is executed _after_ the if statement
-// is executed.
+// Test forward references handled with flow typing
+// Currently, the flow typing will not be applied to definitions forwardly referred.
 class Foo {
 
   def test1(): Unit = {
     val x: String|Null = ???
     if (x == null) return ()
-    val y = x.length // ok: x: String inferred
+    def y = x.length // ok: x: String inferred
     ()
   }
 
+  // This test is similar to test1, but a forward DefDef referring
+  // to y is added after x. y is completed before knowing the
+  // fact "x != null", hence, the type x: String|Null is used.
   def test2(): Unit = {
     val x: String|Null = ???
+    def z = y
     if (x == null) return ()
-    def y = x.length // error: x: String|Null inferred
+    def y = x.length // error: x: String|Null is inferred
     ()
   }
 
+  // Since y is referred before definition, flow typing is not used here.
   def test3(): Unit = {
     val x: String|Null = ???
+    lazy val z = y
     if (x == null) return ()
-    lazy val y = x.length // error: x: String|Null inferred
+    lazy val y = x.length // error: x: String|Null is inferred
     ()
   }
 
+  // This case is not valid but the test1 above is, because
+  // non-lazy value definitions exist between forward references.
```

Comment:
I don't understand what "non-lazy value definitions" means? Do you mean to say that the implicit search forces the completion of `implicit val y`?

Reply:
I mean this kind of forward reference is not allowed. Between forward referenced statements, the value definitions have to be lazy.

Reply:
Ok, can we say
"This case is invalid because z has an implicit forward reference to `y`, but `y` isn't lazy (only forward references to lazy vals are allowed)".